### DISCUSSION TITLE
Regarding instructor assignment preferences in the cpsolver

### DISCUSSION THREAD

**Discussion by MohammedAlmatrafi** [UniTime/unitime#169] [Q&A] [answered]
Hello,
We were using UniTime and recently we switched to the [cpsolver](https://github.com/UniTime/cpsolver) in our project in order to create a customized university course timetables programmatically.

The solver is running perfectly fine but we are missing a feature that is present in UniTime, which is the [Instructor Assignment Preferences](https://help.unitime.org/instructor-assignment-preferences) scheduling options.

This left us asking `if there is a way to implement instructor time preferences using the cpsolver?` because in the [data format](https://www.unitime.org/uct_dataformat_v24.php) we couldn't say it for sure.

Also We were wondering `if the cpsolver is used in the UniTime system` as there are much more features in the system compared to the cpsolver.

Finally We appreciate everyone involved in the development of this project. Thanks.

**Comment by tomas-muller** [comment, marked as answer]
UniTime is using the cpsolver. Moreover, on the Course Timetling Solver page (as well as the solver pages for other problems), you can export the solver that is loaded into memory in XML format directly for the cpsolver. So, you can run the same solver from the command line (just using the cpsolver library) and even import it back to UniTime once it is done. See [help.unitime.org/exporting-solver-xml](https://help.unitime.org/exporting-solver-xml) for more details.

However, when the data are being loaded into the solver, the course timetabling structure and all the preferences and requirements are computed and transformed into the solver's constraint model. In particular, for each class, all available times and room placements are computed, together with their preferences. This is done in the [TimetableDatabaseLoader](https://github.com/UniTime/unitime/blob/master/JavaSource/org/unitime/timetable/solver/TimetableDatabaseLoader.java) class.

You can see the same thing happening in the UniTime user interface. For each class, preferences that are set on the scheduling subpart are combined instructor preferences (of the instructor or instructors assigned to the class) and possibly overridden with preferences set directly on the class level. In the end, what you see on the class (e.g., preferences and available rooms on the Class Detail page) is what the solver will use.

Since you have chosen to use just the cpsolver, you would need to do the same thing if you have some variation on UniTime's instructor preferences. That is, as you are loading the data into the solver (or creating the XML input file for the cpsolver), you will need to combine preferences and requirements from various levels to give the solver a list of possible times and rooms for each class to use, together with their preferences (or penalizations).

If you are interested, we have recently organized an [International Timetabling Competition 2019](https://www.itc2019.org), using the data collected from UniTime from 10 institutions worldwide. However, some simplifications have been made to the problem model without losing most of the problem's complexity. Other than the cpsolver (see [github.com/tomas-muller/cpsolver-itc2019](https://github.com/tomas-muller/cpsolver-itc2019)), there are two other open-source solvers available.
